### PR TITLE
Document ANSIBLE_METADATA

### PR DESCRIPTION
**Pull request by gundalow**
##### ISSUE TYPE
 - Docs Pull Request

##### SUMMARY
Document what was agreed in https://github.com/ansible/proposals/issues/30

---
This change is [Reviewable](https://reviewable.io/reviews/ansible/ansible/21250)

### CODE REVIEW

**Comment by gundalow** [comment]
Needs technical wording improvements from @abadger

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: The rationale here should be:

This is because the ANSIBLE_METADATA through RETURNS sections are not used by the module code itself. They are essentially extra docstrings for the file.  The imports are placed after these special variables for the same reason as PEP8 puts the imports after the introductory comments and docstrings: It keeps the active parts of the code together and the pieces which are purely informational apart.

Comment (abadger) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: The historical note might be more of a warning:  If you look at pre-existing modules, you may find imports at the bottom of the file.  Do not copy that idiom into new modules as it is a historical oddity due to how modules used to be combined with libraries.  Over time we're moving the imports to be in their proper place.

Comment (sivel) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Just to expand on this, some excerpts from my proposal:

> We have decided to continue to ignore E402 module level import not at top of file due to our preference to place imports in modules below the DOCUMENTATION/RETURN/EXAMPLES/METADATA variables, as we treat those similarly to module level docstrings. This allows us to keep all code and imports together instead of split from each other.

> The decision to exclude E402 is based on readability. As described above, documentation strings in a module are much more similar to module level docstrings, than code, and are never utilized by the module itself. Placing the imports below this documentation and closer to the code, consolidates and groups all related code in a congruent manner to improve readability, debugging and understanding.

Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Thanks guys. I'll flesh this out a bit.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Need some into section here

Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: ANSIBLE_METADATA contains information about the module for use by other tools.  At the moment, it informs other tools about what class of people maintains the module and to what degree users can rely on a module's behaviour remaining the same over time.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Note, formatting of this section currently seems to be a quick port from the markdown (since it was in a github issue).  Probably should use rst section headers and field lists isntead.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Could use a warning here. (same structure but use ```.. warning::``` to introduce it)

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:21]: I'd make these links to the relevant sections below.

For linking in general we need to think about how we want to use ref: http://www.sphinx-doc.org/en/stable/markup/inline.html#ref-role  Currently our ref targets are all thrown into the same namespace.  The documentation has gotten big enough that we should think of splitting that a bit.  Probably having a set of high level categories and  prefixing the category name to the ref makes sense.  (That way we can have multiple references to an idea like:  cfg_pipelining, plugin_connection_pipelining, intro_pipelining)  Maybe dev_guide is a sufficient prefix or perhaps dev_mods?

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Possessive, module's

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Good idea, I've done the same for promoting a module

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Correct, I wanted to check people were roughly happy with the content

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (ryansb) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: :+1:

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:21]: Agreed, though I'm going to make this a problem for the future (and I'd want to get Scott's input)

**Comment by gundalow** [comment]
ready_for_review

* Fix headings
* Used field lists (didn't know about this, lot more readable)
* Added in a few notes about empty `options` & `returns`

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: "...module code itself; they are essentially..."

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: add a "?"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: suggest changing "class of people" to something like "type of maintainer"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: comma after "For new modules"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: comma after "release"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: capitalize "maintained"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Suggest "This module is maintained by the community at large, which is responsible for fixing bugs, adding new features, and reviewing changes."

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:87]: Comma after "merging"

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: "There", not "their".

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Change to "If your module doesn't doesn't have any options (for example, if it's a ``_facts`` module), you can use..."

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (dharmabumstead) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: s/you/your; consider adding a comma after the parenthetical

**Review by dharmabumstead — CHANGES_REQUESTED** [review]
Looks pretty good. A few edits.

**Comment by gundalow** [comment]
Willthames made the good suggestion I should document:
`from module_utils.basic import *` 
To
`from module_utils.basic  import AnsibleModule`

Although not documentation related we do refer to imports in here, and currently we don't have anywhere else to add this. 

Once these docs are live (with anchors) `validate-modules` needs updating to point to the online docs.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:21]: I've added a TOC in now, so that should make it easier for people to navigate the page

**Comment by gundalow** [comment]
@willthames I've added in a note about imports, thanks for the suggestion

ready_for_review

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (sivel) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Missing `w` from `ildcard` :)

Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Bonus points for spotting the deliberate mistake to ensure someone had reviewed it.

**Comment by gundalow** [comment]
bot_status

**Comment by gundalow** [comment]
!needs_revision

**Comment by gundalow** [comment]
ready_for_review

**Comment by willthames** [comment]
shipit

**Review by dharmabumstead — APPROVED** [review]
Splendid. Thanks @gundalow!

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: When I see multiple ```.. note::``` adjacent to each other, I wonder if they should be a single ```.. note::``` with a list inside.

(Also:  s/If you module doesn't doesn't/If the module doesn't/)

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Do we want this in a ```.. warning::```?

**Comment by abadger** [comment]
Technical review:  +1 ; saw a pair of formatting and typos.  Not blockers but noted so you can think about them/fix them.

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Excellent idea

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (gundalow) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst]: Converted to list
Change to "If the modules...

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:15]: do we really need to enforce order?

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:62]: so this is 3rd version i've seen, original was 'curated', modules currently use 'committer' and now we are adding 'core_curated'?

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:83]: new value different from example above

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:85]: this is what modules currently use, differs from 'core_curated' in example above

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_modules_documenting.rst:130]: assumes python, any extension could be removed .rb in case of ruby, .pl in case of perl

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (sivel) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:62]: Not that it is necessarily the source of truth, but validate-modules sets the schema to this field as:

```
        Required('supported_by'): Any('core', 'community', 'unmaintained',
                                      'committer')
```

This was done based off of the proposal, but I also see that the metadata tool lists:

core, curated, committer, community

Comment (bcoca) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:62]: weeee!

Comment (bcoca) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:62]: curated == commiter

File: docs/docsite/rst/dev_guide/developing_modules_documenting.rst
Comment (bcoca) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:130]: `The name of the module, This must be the same as it would be used from a play.`

Comment (abadger) [reply docs/docsite/rst/dev_guide/developing_modules_documenting.rst:130]: We have the problem of conflating audiences... I think the audience for this doc is modules that will be accepted into ansible/ansible (which others can look at for best practices but can vary as fits their needs.)

**Pull request merged**
